### PR TITLE
Handle membership in the client on split brain

### PR DESCRIPTION
**Pull request by sancar**
It could be the case that when split brain occurs the member list
version that client remembers is already bigger than the the
half that client is connected to. In this case, client can not
apply the member list because it has an older version. But
in fact, it should have been applied, otherwise client get
stuck with an old member-list.

As a solution, we clear the member-list version everytime
we re-register the member list listener.

fixes https://github.com/hazelcast/hazelcast/issues/16855

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java [hunk 1]
```diff
@@ -324,4 +328,63 @@ public boolean decodeRemoveResponse(ClientMessage clientMessage) {
...
+        blockCommunicationBetween(instanceA, instanceC);
+        closeConnectionBetween(instanceA, instanceC);
+        blockCommunicationBetween(instanceB, instanceC);
+        closeConnectionBetween(instanceB, instanceC);
+
+        assertOpenEventually(" A B | C", splitLatch);
+        instanceB.shutdown();
+        assertOpenEventually(" A | C", bRemovedLatch);
+        instanceA.shutdown();
+
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java:387]: Minor: we don't check whether or not `aRemovedLatch` is open

Comment (sancar) [reply hazelcast/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java:387]: Good catch 👍

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1305/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1305/Hazelcast-pr-builder-1305.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/internal/diagnostics/Diagnostics.java:223: Line has trailing spaces. [RegexpSingleline]
--------------------------
</details>
</pre>

**Comment by sancar** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1321/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1321/Hazelcast-pr-builder-1321.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientQueryCacheRecreationTest.query_cache_recreates_itself_after_server_restart:112->HazelcastTestSupport.assertTrueEventually:1344->HazelcastTestSupport.assertTrueEventually:1242->lambda$query_cache_recreates_itself_after_server_restart$0:114 expected:<200> but was:<185>
[INFO] 
[ERROR] Tests run: 27794, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Pull request merged**